inmaputil/download.go

```diff
@@ -0,0 +1,123 @@
+package inmaputil
+
+import (	
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/mholt/archiver"
+	"github.com/cavaliercoder/grab"
+)
+
+// download checks if the input is an existing file locally.
+// If not, it checks if the file is a URL.
+// If it's a URL, it downloads the file and decompresses it if needed.
+// Then it returns the path to the Shp file.
+// c, if not nil, is a channel across which status update messages will be sent.
+func download(path string, c chan string) string {
+	// Check if local file exists. If it does, return the given path.
+	if _, err := os.Stat(path); !os.IsNotExist(err) {
+		return path
+	} 
+
+	// If the input doesn't start with these prefixes, then return the input.
+	if !strings.HasPrefix(path, "http://") && !strings.HasPrefix(path, "https://") {
+		return path
+	}
+
+	// Prepare a temporary directory for the downloads.
+	tempDir, err := ioutil.TempDir("","inmap");
+	if err != nil {
+		panic(fmt.Errorf("inmaputil: failed creating temporary download directory: %v", err))
+		return path
+	}
+	
+	// Start the downloading client.
+	client := grab.NewClient()
+	req, err := grab.NewRequest(tempDir, path)
+	if err != nil {
+		panic(fmt.Errorf("Failed creating downloading Request."))
+	}
+
+	c <- fmt.Sprintf("Downloading ... %v\n" , req.URL())
+	resp := client.Do(req)
+
+	// Start a ticker to print out progress.
+	t := time.NewTicker(1000 * time.Millisecond)
+	defer t.Stop()
+
+Loop:
+	for {
+		select {
+			case <-t.C:
+				c <- fmt.Sprintf("   transferred %v / %v bytes", resp.BytesComplete(), resp.Size)
+			case <- resp.Done:
+				break Loop
+		}
+	}
+
+	if err := resp.Err(); err != nil {
+		c <- fmt.Sprintf("Download failed: %v", err)
+		return path
+	}
+
+	c <- fmt.Sprintf("Download saved to ./%v ", resp.Filename)
+
+	// Check if the downloaded file is compressed.
+	// If so, decompress it according to its extension.
+	switch filepath.Ext(resp.Filename) {
+	case ".zip":
+		c <- ("Decompressing...")
+		err := archiver.Zip.Open(resp.Filename, resp.Filename + "_extracted")
+		if err != nil {
+			return resp.Filename
```

Comment:
I think for this error and the similar ones below we should also send a message saying what the error is.